inc/renderer.php

```diff
@@ -622,4 +458,198 @@ function enqueue_styles() {
 			$this->add_inline_css( get_the_ID(), $this->generate_css( get_the_ID() ) );
 		}
 	}
+
+	private function get_panels_data_for_post( $post_id ) {
+		if ( strpos( $post_id, 'prebuilt:' ) === 0 ) {
+			list( $null, $prebuilt_id ) = explode( ':', $post_id, 2 );
+			$layouts     = apply_filters( 'siteorigin_panels_prebuilt_layouts', array() );
+			$panels_data = ! empty( $layouts[ $prebuilt_id ] ) ? $layouts[ $prebuilt_id ] : array();
+		} else if ( $post_id == 'home' ) {
+			$page_id = get_option( 'page_on_front' );
+			if ( empty( $page_id ) ) {
+				$page_id = get_option( 'siteorigin_panels_home_page_id' );
+			}
+
+			$panels_data = ! empty( $page_id ) ? get_post_meta( $page_id, 'panels_data', true ) : null;
+
+			if ( is_null( $panels_data ) ) {
+				// Load the default layout
+				$layouts     = apply_filters( 'siteorigin_panels_prebuilt_layouts', array() );
+				$prebuilt_id = siteorigin_panels_setting( 'home-page-default' ) ? siteorigin_panels_setting( 'home-page-default' ) : 'home';
+
+				$panels_data = ! empty( $layouts[ $prebuilt_id ] ) ? $layouts[ $prebuilt_id ] : current( $layouts );
+			}
+		} else {
+			if ( post_password_required( $post_id ) ) {
+				return false;
+			}
+			$panels_data = get_post_meta( $post_id, 'panels_data', true );
+		}
+
+		return $panels_data;
+	}
+
+	private function migrate_widgets_data( $panels_widgets ) {
+
+		// We need this to migrate from the old $panels_data that put widget meta into the "info" key instead of "panels_info"
+		if ( ! empty( $panels_widgets ) && is_array( $panels_widgets ) ) {
+			foreach ( $panels_widgets as $i => $widget ) {
+				if ( empty( $panels_widgets[ $i ]['panels_info'] ) ) {
+					$panels_widgets[ $i ]['panels_info'] = $panels_widgets[ $i ]['info'];
+					unset( $panels_widgets[ $i ]['info'] );
+				}
+
+				$panels_widgets[ $i ]['panels_info']['widget_index'] = $i;
+			}
+		}
+
+		return $panels_widgets;
+	}
+
+	private function get_panels_layout_data( $panels_data ) {
```

Comment:
A phpdoc comment for this function would be useful. You know, so we don't lose it again.